src/tools/tidy/src/no_merge.rs

```diff
@@ -0,0 +1,77 @@
+//! This check makes sure that no accidental merge commits are introduced to the repository.
+//! It forbids all merge commits that are not caused by rollups/bors or subtree syncs.
+
+use std::process::Command;
+
+pub fn check(_: (), bad: &mut bool) {
+    let Ok(remote) = get_rust_lang_rust_remote() else { return; };
+    let Ok(merge_commits) = find_merge_commits(&remote) else { return; };
```

Comment:
I am not a fan of the silent return here. I think a better way of achieving the same aims here is to split this check out, either into a separate tool or gate it on an environment variable, and explicitly only intend that it is run *by default* in CI where we control the environment. And when run, it shouldn't ignore failures to retrieve things like the remote name etc, but rather error.

Reply:
The whole point behind this is to run locally before pushing anything, so it should run by default. My reasoning behind the silent return was that people with bespoke git setups are generally less likely to make such errors so the check isn't that important for them. But I agree that in CI it should error if there's anything wrong. What would the best way be to distinguish between CI and not-CI?

Reply:
I'd be surprised if it's that common to run tidy before pushing -- it seems like a relatively unusual thing to me, but I could be surprised.

https://github.com/rust-lang/rust/blob/master/src/bootstrap/util.rs#L260

Reply:
@Mark-Simulacrum we run tidy in the pre-push hook `x.py setup ` creates

Reply:
Hm, okay. I guess it seems like if you're running via pre-push hook, then you at least have git, right? So presumably in that case we can maybe pass something from there...

In any case, I think at minimum actually running it in CI seems like a good idea.